src/home/homework-blob.c

```diff
@@ -0,0 +1,236 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * Copyright © 2024 GNOME Foundation Inc.
+ *      Original Author: Adrian Vovk
+ */
+
+#include "copy.h"
+#include "fileio.h"
+#include "fd-util.h"
+#include "fs-util.h"
+#include "home-util.h"
+#include "homework-blob.h"
+#include "homework.h"
+#include "install-file.h"
+#include "macro.h"
+#include "path-util.h"
+#include "recurse-dir.h"
+#include "rm-rf.h"
+#include "sha256.h"
+#include "string-util.h"
+#include "tmpfile-util.h"
+#include "umask-util.h"
+#include "utf8.h"
+
+#define BLOB_MAX_SIZE (64U*1024U*1024U)
+
+static int copy_one_blob(
+                int src,
+                int dest_dfd,
+                const char *name,
+                uint64_t *total_size,
+                uid_t uid,
+                Hashmap *manifest) {
+        _cleanup_close_ int dest = -EBADF;
+        uint8_t hash[SHA256_DIGEST_SIZE];
+        off_t initial, size;
+        int r;
+
+        if (!hashmap_contains(manifest, name)) {
+                log_warning("File %s in blob directory is missing from manifest. Skipping.", name);
+                return 0;
+        }
+
+        if (!suitable_blob_filename(name)) {
+                log_warning("File %s in blob directory has invalid filename. Skipping.", name);
+                return 0;
+        }
+
+        initial = lseek(src, 0, SEEK_CUR);
+        if (initial < 0)
+                return log_debug_errno(errno, "Failed to get initial pos on fd for %s in blob: %m", name);
+
+        r = sha256_fd(src, hash);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to compute sha256 for %s in blob: %m", name);
+
+        size = lseek(src, 0, SEEK_CUR);
+        if (size < 0)
+                return log_debug_errno(errno, "Failed to get final pos on fd for %s in blob: %m", name);
+        size -= initial;
+
+        if (lseek(src, initial, SEEK_SET) < 0)
+                return log_debug_errno(errno, "Failed to rewind fd for %s in blob: %m", name);
+
+        if (memcmp(hash, hashmap_get(manifest, name), SHA256_DIGEST_SIZE) != 0) {
+                log_warning("File %s in blob directory has incorrect hash. Skipping.", name);
+                return 0;
+        }
+
+        *total_size += size;
```

Comment:
i am in that age where any addition like this (i.e. user controlled sizes) makes me worred. 

i.e. can you add an overflow check first:

```c
if (*total_size > UINT64_MAX - size) … 
```

if that passes an overflow can't happen.

Note that an overflow is easy to trigger here: people can generate sparse files which look immense but aren't.

Reply:
I added such a check, but there are [some compiler builtins](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html) for this which use dedicated CPU instructions if they exist. Should we use them, maybe w/ a nicer name `#define`d for them?

For instance, this:

```c
if (_likely_(*total_size <= UINT64_MAX - size))
    *total_size += size;
else
    *total_size = UINT64_MAX;
```

can become this:

```c
if (!__builtin_add_overflow(*total_size, size, total_size))
    *total_size = UINT64_MAX;
```

Reply:
I added helper macros for the compiler builtins and used them

Reply:
some people are fans of those builtins, but i find them hard to read, i never know what it actually does. also, i am pretty sure the compiler doesn't need that, it's fully capable of recognizing overflow checks on its own, and making them smarter

Reply:
but i guess if you add the macros you are proposing, then it's ok. i figure this will come up over and over again, and i do must admit that the typesafety of the builtins is kinda nice.